devices/src/virtio/vsock/unix/muxer.rs

```diff
@@ -0,0 +1,736 @@
+                        local_port: pkt.dst_port(),
+                        peer_port: pkt.src_port(),
+                    })
+                    .unwrap_or_else(|_| {
+                        info!("vsock: muxer.rxq full - unable to enqueue RST for peer");
+                    });
+            });
+    }
+
+    /// Perform an action that might mutate a connection's state.
+    ///
+    /// This is used as shorthand for repetitive tasks that need to be performed after a connection
+    /// object mutates. E.g.
+    /// - update the connection's epoll listener;
+    /// - schedule the connection to be queried for RX data;
+    /// - kill the connection if an unrecoverable error occurs.
+    ///
+    fn apply_conn_mutation<F>(&mut self, key: ConnMapKey, mut_fn: F)
+    where
+        F: FnOnce(&mut MuxerConnection),
+    {
+        if let Some(conn) = self.conn_map.get_mut(&key) {
+            let had_rx = conn.has_pending_rx();
+            let was_expiring = conn.will_expire();
+
+            mut_fn(conn);
+
+            // If the connection wasn't previously scheduled for RX, add it to our RX queue.
+            if !had_rx && conn.has_pending_rx() {
+                self.rxq.push(MuxerRx::ConnRx(key)).unwrap_or_default();
+            }
+
+            // If the connection wasn't previously scheduled for termination, add it to the kill queue.
+            if !was_expiring && conn.will_expire() {
+                self.killq.push(key, conn.expiry().unwrap());
+            }
+
+            let fd = conn.get_polled_fd();
+            let new_evset = conn.get_polled_evset();
+            if new_evset.is_empty() {
+                // If the connection no longer needs epoll notifications, remove its listener from
+                // our list.
+                self.remove_listener(fd);
+                return;
+            }
+            if let Some(EpollListener::Connection { evset, .. }) = self.listener_map.get_mut(&fd) {
+                if *evset != new_evset {
+                    // If the set of events that the connection is interested in has changed, we need
+                    // to update its epoll listener.
+                    debug!(
+                        "vsock: updating listener for (lp={}, pp={}): old={:?}, new={:?}",
+                        key.local_port, key.peer_port, *evset, new_evset
+                    );
+
+                    *evset = new_evset;
+                    epoll::ctl(
+                        self.epoll_fd,
+                        epoll::ControlOptions::EPOLL_CTL_MOD,
+                        fd,
+                        epoll::Event::new(new_evset, fd as u64),
+                    )
+                    .unwrap_or_else(|err| {
+                        // This really shouldn't happen, like, ever. However, "famous last words" and
+                        // all that, so let's just kill it with fire, and walk away.
+                        self.kill_connection(key);
+                        warn!(
```

Comment:
Yeah, I kind of agree, because `epoll_ctl` shouldn't fail. Bumped.

However, note my exception classification comment in #1106. The gist of it being that, even though this is an error for the muxer, is it still an error in the grand scheme of all things Firecracker? I'm a bit reluctant to pollute the log with a boy-who-cried-wolf approach.